### DISCUSSION TITLE
How to parse traits from error reason?

### DISCUSSION THREAD

**Discussion by mahenzon** [ory/kratos#1713] [Q&A] [answered]
### TL;DR
How to parse "reason" field with traits details? Is there a way to get json with nested fields?

How can this be parsed to show invalid feedback in forms?
```
I[#/traits] S[#/properties/traits/required] missing properties: "username"
```
or
```
I[#/traits] S[#/properties/traits] validation failed
  I[#/traits] S[#/properties/traits/required] missing properties: "email"
  I[#/traits/username] S[#/properties/traits/properties/username] validation failed
    I[#/traits/username] S[#/properties/traits/properties/username/minLength] length must be >= 3, but got 0
    I[#/traits/username] S[#/properties/traits/properties/username/pattern] does not match pattern "^[a-zA-Z]([-_a-zA-Z0-9]{2,19})+$"
```

### More detailed example:
For example I want to update identity, but I'm missing some fields (email and username):

```http
PUT http://127.0.0.1:4434/identities/86f91787-87d5-48a6-9740-aa8dba4750cd

{
 "state": "active",
  "traits": {
    "name": {
      "last": "John",
      "first": "Smith"
    }
  }
}
```

then I get error 400 bad request and this json:

```json
{
  "error": {
    "code": 400,
    "status": "Bad Request",
    "reason": "I[#/traits] S[#/properties/traits/required] missing properties: \"email\", \"username\"",
    "message": "The request was malformed or contained invalid parameters"
  }
}
```

Or I'm violating some rules:

```http
PUT http://127.0.0.1:4434/identities/86f91787-87d5-48a6-9740-aa8dba4750cd

{
 "state": "active",
  "traits": {
    "name": {
      "last": "John",
      "first": "Smith"
    },
    "username": ""
  }
}
```

And I get

```json
{
  "error": {
    "code": 400,
    "status": "Bad Request",
    "reason": "I[#/traits] S[#/properties/traits] validation failed\n  I[#/traits] S[#/properties/traits/required] missing properties: \"email\"\n  I[#/traits/username] S[#/properties/traits/properties/username] validation failed\n    I[#/traits/username] S[#/properties/traits/properties/username/minLength] length must be >= 3, but got 0\n    I[#/traits/username] S[#/properties/traits/properties/username/pattern] does not match pattern \"^[a-zA-Z]([-_a-zA-Z0-9]{2,19})+$\"",
    "message": "The request was malformed or contained invalid parameters"
  }
}
```

Thanks!

**Comment by aeneasr** [comment, marked as answer]
That is indeed currently not quite possible for admin api interaction! Could you share your use case why you want it parsed?

**Reply by mahenzon** [reply]
User is updating identity details through our API, our API sends a request to kratos admin API to update details, response is returned to client. So if there's an error, we want to add invalid feedback to the form fields.

**Reply by aeneasr** [reply]
I see, however I really discourage doing it that way due to security issues (account enumeration, account takeover). Why are you not using the settings flow for this scenario?

However you are in bit of luck, there is a new version of the JSON Validator which came out today. But it will take a few months before it lands here probably as we need to update the fork and do some other things before :/ If you are proficient in Go you could gibe it a try!

**Reply by mahenzon** [reply]
If this is a wrong way, what's the right way to update identity details (user's name and extra fields)?

**Reply by aeneasr** [reply]
Using the settings flow! See docs for more info :)

https://www.ory.sh/kratos/docs/self-service/flows/user-settings
